### PR TITLE
Replace ring buffer with moodycamel/concurrentqueue

### LINKED ISSUE #96: Replace ring buffer implementation
We can remove a chunk of code from code that we have to maintain, and there _may_ be performance to gain from switching to a lock-free implementation.
The following implementation seems to be more highly regarded than the options available in boost:
https://github.com/cameron314/concurrentqueue

Benchmarking:
https://github.com/Qihoo360/evpp/blob/master/docs/benchmark_lockfree_vs_mutex.md#brief

### PR DESCRIPTION
**Pull request by rerpha**
### Description of work

Replaced the ring buffer in `Ring.h/cpp` used for epics PV updates and conversion workers with a [moodycamel concurrentqueue](https://github.com/cameron314/concurrentqueue) buffer 

### Issue

Closes #96 

### Acceptance Criteria

No change in functionality although slight difference in how the ring buffer is used as concurrentqueue is a lockless buffer so some mutexes have been removed. 

### Unit Tests

No unit tests modified

### Other

contains the changes in #104 so please don't merge before 104 is merged. 

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

File: src/Stream.cpp [hunk 1]
```diff
@@ -79,46 +75,44 @@ int Stream::converter_add(InstanceSet &kset, Converter::sptr conv,
...
+    std::unique_ptr<FlatBufs::EpicsPVUpdate> e;
+    auto found = emit_queue->try_dequeue(e);
     n0 += 1;
-    if (e.first != 0) {
+    if (!found) {
       CLOG(6, 1, "empty? should not happen");
       break;
     }
-    auto &up = e.second;
+    auto &up = e;
```
Comment (mattclarke) [inline src/Stream.cpp]: I don't think we need &up anymore as we can use e directly.

File: src/Stream.cpp [hunk 2]
```diff
@@ -79,46 +75,44 @@ int Stream::converter_add(InstanceSet &kset, Converter::sptr conv,
...
     if (!up) {
       LOG(6, "empty epics update");
       continue;
     }
     size_t cpid = 0;
     uint32_t ncp = conversion_paths.size();
-    on_seq_data(e.second->seq_data);
+    on_seq_data(e->seq_data);
     for (auto &cp : conversion_paths) {
       auto p = std::unique_ptr<ConversionWorkPacket>(new ConversionWorkPacket);
```
Comment (mattclarke) [inline src/Stream.cpp]: make_unique?

File: src/Stream.cpp [hunk 3]
```diff
@@ -79,46 +75,44 @@ int Stream::converter_add(InstanceSet &kset, Converter::sptr conv,
...
         p->up = std::unique_ptr<FlatBufs::EpicsPVUpdate>(
-            new FlatBufs::EpicsPVUpdate(*e.second));
+            new FlatBufs::EpicsPVUpdate(*e));
       }
-      auto x = q2.push_unsafe(p);
-      if (x != 0) {
+      bool x = q2.enqueue(std::move(p));
+      if (x) {
         CLOG(6, 1, "full? should not happen");
         break;
```
Comment (mattclarke) [inline src/Stream.cpp]: Can this log message (and the ones above) have more detail.
E.g. `Conversion work queue is full`

File: src/EpicsClient/EpicsClientMonitor.cpp [hunk 4]
```diff
@@ -166,7 +168,7 @@ int EpicsClientMonitor::emit(std::unique_ptr<FlatBufs::EpicsPVUpdate> up) {
     // should never happen, ignore
     return 0;
   }
-  emit_queue->push_enlarge(up);
+  emit_queue->enqueue(std::move(up));
   return 1;
 }
```
Comment (rerpha) [inline src/EpicsClient/EpicsClientMonitor.cpp]: returns 1 on success?

File: src/Stream.cpp [hunk 5]
```diff
@@ -71,58 +67,53 @@ int Stream::converter_add(InstanceSet &kset, Converter::sptr conv,
                           URI uri_kafka_output) {
   auto pt = kset.producer_topic(uri_kafka_output);
   std::unique_ptr<ConversionPath> cp(new ConversionPath(
```
Comment (rerpha) [inline src/Stream.cpp]: use make_unique

File: src/tests/streams_tests.cpp [hunk 6]
```diff
@@ -20,9 +19,10 @@ class FakeEpicsClient : public EpicsClient::EpicsClientInterface {
 
 std::shared_ptr<Stream> createStream(std::string provider_type,
                                      std::string channel_name) {
-  auto ring =
-      std::make_shared<Ring<std::unique_ptr<FlatBufs::EpicsPVUpdate>>>();
-  auto client = ::make_unique<FakeEpicsClient>();
+  auto ring = std::make_shared<
+      moodycamel::ConcurrentQueue<std::unique_ptr<FlatBufs::EpicsPVUpdate>>>();
+  auto client =
+      std::unique_ptr<EpicsClient::EpicsClientInterface>(new FakeEpicsClient());
```
Comment (rerpha) [inline src/tests/streams_tests.cpp]: use make_unique

**Pull request merged**